### ISSUE TITLE
Validation step metrics not logged

### ISSUE THREAD

**Issue by feribg** [PyTorchLightning/pytorch-lightning#2102]
## ❓ Questions and Help

It seems like data output in the `validation_step` does not get logged to tensorboard, it needs to be aggregated first in the `validation_epoch_end`, which is not the case for `training_step`.

The below would only show `val_loss` and only aggregated but all of `mae`, `mape` etc from every iteration.

As a workaround I could explicitly log, but how do I get the current iteration in the callbacks I only see how to get the current epoch.

```
    def step(self, y_hat, y, mode='train'):
        loss = F.mse_loss(y_hat, y)
        mae = F.l1_loss(y_hat, y)
        mape = median_absolute_percentage_error(y_hat, y)
        r2 = r2_score(y_hat, y)
        out = {'loss': loss, 'mae': mae, 'mape': mape, 'R2': r2}
        if mode=='train':
            out['log'] = out.copy()
            return out
        elif mode =='val':
            out = {f'{mode}_{k}': v for k,v in out.items()}
            out['log'] = out.copy()
            return out
        else:
            raise Exception('Unsupported mode')
    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        return self.step(y_hat, y, 'val')

    def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        tensorboard_logs = {'val_loss': avg_loss}
        return {'val_loss': avg_loss, 'log': tensorboard_logs}
```

**Comment by awaelchli** [comment]
you can get the current step with `self.global_step` or self.trainer.global_step

yes, it's true you can manually log it yourself, but there is a reason why we don't log each step of the validation, because the loggers (all as far as I know) use a global step for logging, and this means if your training epoch has n bathces, and your validation has m batches, after the first epoch you will log
n + m steps for training + validation and then your training loss will contingue with step n+m+1 instead of n+1 for epoch 2. You will see a  big jump. 

TensorBoard is too limited, you cannot set the abscissa to anything else than the step. Therefore logging validation step makes no sense.

**Comment by awaelchli** [comment]
duplicate of #1906

**Comment by feribg** [comment]
To your point re the step # being inconsistent, I'm not sure why that would be the case when the `val` metrics are always prefixed with `val_` so you will get `m` number of `val_...` metric for each validatio epoch, right? I'm struggling to understand where the mixup between train and valid comes into play when they are namespaced differently  and stored as different metrics?

**Comment by awaelchli** [comment]
epoch 1 starts
batch 1: train_loss = ... gloabal_step = 1
batch 2: train_loss = ... gloabal_step = 2
batch 3: train_loss = ... gloabal_step = 3
validation starts
batch 1: val_loss = ... gloabal_step = 4
batch 2: val_loss = ... gloabal_step = 5
epoch 2 starts
batch 1: train_loss = ... **gloabal_step =6**
batch 2: train_loss = ... **gloabal_step = 7**
batch 3: train_loss = ... **gloabal_step = 8**

try it yourself :)

this is a limitation of the loggers, which cannot log to the past history, only at the current global step.

**Comment by awaelchli** [comment]
why do you need to log every validation step anyway? it's not going to tell you much about your validation results

**Comment by feribg** [comment]
You're right I don't really need to, just wanted to keep the plots on the same axis and the code consistent between train / valid. Feel free to close it if you don't think it's worthwhile.
